### PR TITLE
[v2] gatsby develop - prefetch data and wait for it before changing path

### LINKED ISSUE #1625: [gatsby-plugin-twitter] Not consistently working
Excuse the vague issue title, haven't had the time to look into the code actually, so all I can do for now is report what i observed: On gatsbyjs.org, navigating from the homepage to "Blog", then to "Announcing Gatsby 1.0.0", tweets are rendering fine. When going back to "Blog", and then accessing that article again, things aren't working any more:

image

Adblockers and related stuff that might block Twitter's widgets are off. I suppose this is an issue with gatsby-plugin-twitter because on https://www.yisela.com/computer-generated-poetry/ I observe a similar behavior – only tweets do not render more often than they do over there, and I can not really reproduce a certain behavior like on gatsbyjs.org.

### LINKED ISSUE #5094: [v2] In development, fetch page query data when link component mounts to avoid delay loading data

### LINKED ISSUE #6309: Anchor links don't work in develop
I have an element with the id `hash` on [my page](https://github.com/fk/gatsby-v2-anchor-link/blob/6d886ed4025a14f3e8605a9ed26e6331d7887f61/src/pages/index.js#L22):

```
<h1 id="hash">Headline with #hash</h1>
```

When running `gatsby develop`, opening a new browser window and accessing `http://localhost:8000/#hash`, I expect the page to jump to the element with the corresponding id `hash`. It doesn't: 

image

Only when I either
* open just `http://localhost:8000/` and then add `#hash` to the URL in the browser address bar and hit "Enter" or
* run `gatsby build` and then open `http://localhost:9000/#hash`,

the page jumps to the element as expected:

image

Repo here: https://github.com/fk/gatsby-v2-anchor-link

### PR DESCRIPTION
**Pull request by pieh**
- Removes navigation code duplicated in production and dev runtime.
- Waits for page data before transitioning to new page in dev
- Limits page re-renders - only updates JSONStore component if:
  - path changed
  - data for current path changed
  - data for static queries changed

Closes #5094
Closes #6309 
Closes #1625

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit 4349afbba52f336640798d89b66b718a9a3fc6e7

https://app.netlify.com/sites/using-postcss-sass/deploys/5b59d9233813f05f24990982

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit 6258ef4060dddba3ee44bbcdf4350c2c99de33e9

https://app.netlify.com/sites/gatsbyjs/deploys/5b5ad2b402ed8340c790f414

File: packages/gatsby/cache-dir/json-store.js [hunk 1]
```diff
@@ -2,7 +2,18 @@ import React from "react"
...
+import {
+  getStaticQueryData,
+  getPageQueryData,
+  registerPath as socketRegisterPath,
+  unregisterPath as socketUnregisterPath,
+} from "./socketIo"
+
+if (process.env.NODE_ENV === `production`) {
+  throw new Error(
+    `It appears like Gatsby is misconfigured. JSONStore shouldn't be used in production.`
```
Comment (m-allanson) [inline packages/gatsby/cache-dir/json-store.js]: I guess this _should_ never happen? Can the message be updated with advice on what to do if it does happen?

File: packages/gatsby/src/utils/websocket-manager.js [hunk 2]
```diff
@@ -148,20 +151,31 @@ class WebsocketManager {
...
-        this.activePaths.add(path)
-
+      const getDataForPath = path => {
         if (!this.pageResults.has(path)) {
           const result = getCachedPageData(path, this.programDir)
-          this.pageResults.set(path, result)
+          if (result) {
+            this.pageResults.set(path, result)
+          } else {
+            console.log(`Page not found`, path)
```
Comment (m-allanson) [inline packages/gatsby/src/utils/websocket-manager.js]: I'm seeing this error output even though the page is loading.

Clicking around the blog, and then the tags page (http://localhost:8000/blog/tags) and individual tag pages

```
View GraphiQL, an in-browser IDE, to explore your site's data and schema

  http://localhost:8000/___graphql

Note that the development build is not optimized.
To create a production build, use gatsby build

boundActionCreators is deprecated. Use actions instead. Check the following files:

src/pages/docs/actions.js

ℹ ｢wdm｣:
ℹ ｢wdm｣: Compiled successfully.
Page not found /blog/tags/documentation
Page not found /blog/tags/launch
Page not found /blog/tags
Page not found /blog/tags
Page not found /blog/tags
Page not found /blog/tags
```

**Comment by KyleAMathews** [comment]
Deploy preview for *using-jss* failed.

Built with commit 6b4f45ace5c8ac4ac24cdf7fd7e31dfa05d35451

https://app.netlify.com/sites/using-jss/deploys/5b6071403813f06353f8da19

File: packages/gatsby/src/utils/websocket-manager.js (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/utils/websocket-manager.js]: This is fixed by https://github.com/gatsbyjs/gatsby/pull/6777/commits/5fdcc4ca68fe4b50e596d2c480f01b751cdba3b1

File: packages/gatsby/cache-dir/json-store.js (hunk 1 above)
Comment (pieh) [reply packages/gatsby/cache-dir/json-store.js]: I expanded this message in https://github.com/gatsbyjs/gatsby/pull/6777/commits/acfebf50fdeb3e360575d9e574f634e6e7d2d259 , but really only advice there is to report it as issue here

**Comment by pieh** [comment]
I think this is good to go!

**Comment by m-allanson** [comment]
Published as `gatsby@2.0.0-beta.65`

**Pull request merged**